canvas/api/assignment/fetchsubmissionfiles.py

```diff
@@ -100,25 +102,41 @@ def _write_files(temp_dir, file_contents):
     return len(file_contents)
 
 def _download_files(temp_dir, file_targets):
+    num_files = sum([len(targets) for targets in file_targets.values()])
+    if (num_files == 0):
+        return 0
+
     count = 0
+    max_workers = min(MIN_WORKERS , num_files)
+    tasks = []
 
-    for (email, targets) in file_targets.items():
+    for email, targets in file_targets.items():
         user_dir = os.path.join(temp_dir, email)
         os.makedirs(user_dir, exist_ok = True)
 
-        for (filename, url) in targets:
-            path = os.path.join(user_dir, filename)
+        for filename, url in targets:
+            tasks.append((email, filename, url, user_dir))
+
+    try:
+        with concurrent.futures.ThreadPoolExecutor(max_workers = max_workers) as executor:
+            results = executor.map(lambda t: download_file(t[0], t[1], t[2], t[3]), tasks)
+            count = sum(results)
+    except Exception as ex:
+        logging.error("Error while downloading files.", exc_info = ex)
 
-            try:
-                logging.debug("Downloading file for student '%s': '%s'." % (email, filename))
-                response = requests.get(url)
-                response.raise_for_status()
+    return count
 
-                with open(path, 'wb') as file:
-                    file.write(response.content)
+def download_file(email, filename, url, user_dir):
+    path = os.path.join(user_dir, filename)
 
-                count += 1
-            except Exception as ex:
-                logging.error("Failed to write downloaded file ('%s') for student '%s'." % (filename, email), exc_info = ex)
+    try:
+        logging.debug("Downloading file for student '%s': '%s'." % (email, filename))
+        response = requests.get(url, timeout = 10)
```

Comment:
What's with the timeout?

Even if you want it to exist, why is it not a constant?